app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php

```diff
@@ -0,0 +1,506 @@
+                    $leadFieldValue = $leadCompany->{'get'.ucfirst($alias)}();
+                    if (null !== $leadFieldValue) {
+                        $leadCompany->{'set'.ucfirst($alias)}(null);
+                    }
+                    continue;
+                }
+            }
+        }
+
+        return $leadsCompanies;
+    }
+
+    /**
+     * @param array<Lead>    $leads
+     * @param array<Company> $companies
+     *
+     * @return array<int,array<mixed>>
+     */
+    private function setHashFields(array $leads, array $companies, LeadField $field, bool $pseudonymize): array
+    {
+        return [
+            $this->setHashes($leads, $field, $pseudonymize),
+            $this->setHashes($companies, $field, $pseudonymize),
+        ];
+    }
+
+    /**
+     * @param array<Lead>|array<Company> $leadsCompanies
+     *
+     * @return array<mixed>
+     */
+    private function setHashes(array $leadsCompanies, LeadField $field, bool $pseudonymize): array
+    {
+        foreach ($leadsCompanies as $key => $leadCompany) {
+            if (!method_exists($leadCompany, 'getField')) {
+                continue;
+            }
+            $leadField = false;
+            if ($leadCompany instanceof Company && 'company' === $field->getObject()) {
+                $leadField = $leadCompany->getField($field->getAlias());
+                if (false === $leadField) {
+                    $alias          = self::COMPANY_FIELDS_TO_COLUMNS[$field->getAlias()] ?? $field->getAlias();
+                    $leadFieldValue = $leadCompany->getFieldValue($alias);
+                    if (property_exists($leadCompany, $alias) && null !== $leadFieldValue) {
+                        $tempLeadField = $leadCompany->{'get'.ucfirst($alias)}();
+                        if (empty($tempLeadField)) {
+                            continue;
+                        }
+                        unset($leadField);
+                        $leadField['value'] = $tempLeadField;
+                        $leadField['type']  = $field->getType();
+                    } else {
+                        continue;
+                    }
+                }
+            }
+            if ($leadCompany instanceof Lead && 'lead' === $field->getObject()) {
+                $leadField = $leadCompany->getField($field->getAlias());
+                if (false === $leadField) {
+                    continue;
+                }
+
+                $field     = $this->fieldModel->getRepository()->find($leadField['id']);
+
+                if (null === $field) {
+                    continue;
+                }
+            }
+            if (false === $leadField) {
+                continue;
+            }
+
+            $leadsCompanies[$key] = $this->setHash($leadCompany, $leadField, $field, $pseudonymize);
+        }
+
+        return $leadsCompanies;
+    }
+
+    /**
+     * @param array<string, string|null> $field
+     */
+    private function setHash(
```

Comment:
Consider delegating this logic to a service class to improve the separation of concerns.